limits/filetransfer/limits.go

```diff
@@ -0,0 +1,391 @@
+		deviceCountersLastH.bytesTransferred >= p.limits.FileTransfer.Counters.MaxBytesTxPerHour {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesTransferred < math.MaxUint64-n {
+			p.counters.bytesTransferred += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BytesReceived(n uint64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	countersMutex.Lock()
+	defer countersMutex.Unlock()
+
+	belowLimit = true
+	if n != 0 {
+		if deviceCountersLastH.bytesReceived < math.MaxUint64-n {
+			deviceCountersLastH.bytesReceived += n
+		}
+	}
+	if p.limits.FileTransfer.Counters.MaxBytesRxPerHour > 0 &&
+		deviceCountersLastH.bytesReceived >= p.limits.FileTransfer.Counters.MaxBytesRxPerHour {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesReceived < math.MaxUint64-n {
+			p.counters.bytesReceived += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BelowMaxAllowedFileSize(offset int64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	if offset < 0 {
+		return true
+	}
+	if p.limits.FileTransfer.MaxFileSize > 0 &&
+		uint64(offset) >= p.limits.FileTransfer.MaxFileSize {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (p *Permit) PreserveModes(path string, mode os.FileMode) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	if (mode & S_ISUID) != 0 {
+		mode &= os.ModePerm
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode |= os.ModeSetuid
+	} else {
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode &= os.ModePerm
+	}
+
+	if !p.limits.FileTransfer.DoNotPreserveMode {
+		return os.Chmod(path, mode)
+	} else {
+		return nil
+	}
+}
+
+func (p *Permit) PreserveOwnerGroup(path string, uid int, gid int) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	forcedSet := false
+	if p.limits.FileTransfer.OwnerPut != "" {
+		u, err := user.Lookup(p.limits.FileTransfer.OwnerPut)
+		if err != nil {
+			return err
+		}
+		uid, err = strconv.Atoi(u.Uid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if p.limits.FileTransfer.GroupPut != "" {
+		u, err := user.LookupGroup(p.limits.FileTransfer.GroupPut)
+		if err != nil {
+			return err
+		}
+		gid, err = strconv.Atoi(u.Gid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if forcedSet {
+		return os.Chown(path, uid, gid)
+	}
+	if !p.limits.FileTransfer.DoNotPreserveOwner {
+		return os.Chown(path, uid, gid)
+	} else {
+		return nil
+	}
+}
+
+func updatePerHourCounters() {
```

Comment:
ok. let me see if it is easily doable.

Reply:
it would be a bit harder to do with the averaging we do now, since it has to be updated with the const interval (technically does not, but it really simplifies things). so I would keep it async, especially for future task (see below comment on making the limits global for mender-connect all sessions) for keeping a limit for all transfers made by mender-connect.